### PR TITLE
Add AnalysisManager and refactor analysis framework

### PR DESCRIPTION
**Pull request by k-sareen**
I've changed the framework as per the discussion here [1]. Further comments and recommendations are welcome.

~Kunal
\[1]: https://github.com/mmtk/mmtk-core/pull/227#issuecomment-780270882

### CODE REVIEW

File: src/plan/global.rs [hunk 1]
```diff
@@ -437,7 +432,11 @@ impl<VM: VMBinding> BasePlan<VM> {
         constraints: &'static PlanConstraints,
     ) -> BasePlan<VM> {
         let stats = Stats::new();
-        let ctr = stats.new_event_counter("gc.num", true, true);
+        // Initializing the analysis manager and routines
+        #[cfg(feature = "analysis")]
+        let mut analysis_manager = Mutex::new(AnalysisManager::new());
+        #[cfg(feature = "analysis")]
+        analysis_manager.lock().unwrap().initialize_routines(&stats);
```
Comment (qinsoon) [inline src/plan/global.rs:439]: Is there any reason we need to `new()` then `initialize_routines()`? You could make `initialize_routines()` a private function, and call it in `new()`.

File: src/util/analysis/mod.rs [hunk 2]
```diff
@@ -1,22 +1,86 @@
...
+
 pub mod gc_count;
 pub mod obj_num;
 pub mod obj_size;
 
+use self::gc_count::GcCounter;
+use self::obj_num::ObjectCounter;
+use self::obj_size::PerSizeClassObjectCounter;
+
 /**
```
Comment (qinsoon) [inline src/util/analysis/mod.rs:15]: Probably use `///` for rustdoc.

**Review by qinsoon — APPROVED** [review]
Good refactoring. I only have a few minor comments.

**Comment by k-sareen** [comment]
Updated in latest commit.

**Pull request merged**